# RNA-Seq Analysis

Some basic RNA-seq analysis resources.

## Web Resources

[RNA-seqlopedia](http://rnaseq.uoregon.edu/)
> The RNA-seqlopedia provides an overview of RNA-seq and of the choices necessary to carry out a successful RNA-seq experiment.

[Thinking About RNA Seq Experimental Design for Measuring Differential Gene Expression: The Basics](http://gkno2.tumblr.com/post/24629975632/thinking-about-rna-seq-experimental-design-for)
> RNA-Seq is a powerful tool that can be used effectively by a diverse community of people with different backgrounds. We expect that some of those who could benefit from RNA Seq do not yet have the background in sequencing and statistics that is necessary to make effective use of this technology. Much of the existing literature may be time-consuming to read without this background. Therefore, we put together this primer with the intention of helping scientists and students understand the basic statistical principles associated with measuring gene expression using RNA Seq.

[RNA Sequence Analysis Training/Courses/Papers](https://www.biostars.org/p/174376/)
> I did a little bit of research and started to take free online courses and read papers. Could anybody recommend me more up-to-date online trainings or courses? Here are the resources that I found so far.

## Publications

[RNA-Seq workflow: gene-level exploratory analysis and differential expression](http://f1000research.com/articles/4-1070/v1) (10/2015)
> Here we walk through an end-to-end gene-level RNA-Seq differential expression workflow using Bioconductor packages. We will start from the FASTQ files, show how these were aligned to the reference genome, and prepare a count matrix which tallies the number of RNA-seq reads/fragments within each gene for each sample. We will perform exploratory data analysis (EDA) for quality assessment and to explore the relationship between samples, perform differential gene expression analysis, and visually explore the results.

[A survey of best practices for RNA-seq data analysis](http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0881-8) (1/2016)
> We review all of the major steps in RNA-seq data analysis, including experimental design, quality control, read alignment, quantification of gene and transcript levels, visualization, differential gene expression, alternative splicing, functional analysis, gene fusion detection and eQTL mapping. We highlight the challenges associated with each step. We discuss the analysis of small RNAs and the integration of RNA-seq with other functional genomics techniques. Finally, we discuss the outlook for novel technologies that are changing the state of the art in transcriptomics.
